addons/stock_account/models/account_invoice.py

```diff
@@ -77,6 +77,56 @@ def _anglo_saxon_sale_move_lines(self, i_line):
                 ]
         return []
 
+    def _get_related_stock_moves(self):
+        """ To be overridden for customer invoices and vendor bills in order to
+        return the stock moves related to this invoice.
+        """
+        return self.env['stock.move']
+
+    def _get_products_set(self):
+        """ Returns a recordset of the products contained in this invoice's lines
+        """
+        return self.mapped('invoice_line_ids.product_id')
+
+    def _get_anglosaxon_interim_account(self, product):
+        """ Returns the interim account used in anglosaxon accounting for
+        this invoice"""
+        if self.type in ('out_invoice', 'out_refund'):
+            return product.product_tmpl_id._get_product_accounts()['stock_output']
+        elif self.type in ('in_invoice', 'in_refund'):
+            return product.product_tmpl_id.get_product_accounts()['stock_input']
+
+        return None
+
+    def invoice_validate(self):
+        super(AccountInvoice, self).invoice_validate()
+        self.anglo_saxon_reconcile_valuation(correct_invoice_interim_account=True)
+
+    def anglo_saxon_reconcile_valuation(self, correct_invoice_interim_account=False):
+        """ Reconciles the entries made in the interim accounts in anglosaxon accounting,
+        reconciling stock valuation move lines with the invoice's.
+        """
+        for invoice in self:
+            company_currency = invoice.company_id.currency_id
+            if invoice.company_id.anglo_saxon_accounting:
+                invoice_stock_moves = invoice._get_related_stock_moves()
+                all_stock_move_invoices = invoice_stock_moves._get_related_invoices()
```

Comment:
these 2 lines are dubious (and probably the design too, by cascade): 
- when we start from the move, we shouldn't have to look for _get_related_stock_moves()
- when we start from the invoice, we shouldn't have to look for _get_related_invoices()